proposals/libp2p-protocol-select.md

```diff
@@ -0,0 +1,176 @@
+The goal of project proposals is to help us decide which work to take on, which things are more valuable than other things.
+-->
+<!--
+A minimal project pitch (MPP) should contain enough detail for others to understand what problem this project solves and why this is important for our
+team's goal of achieving product-market fit, a high-level description of what the idea/proposed solution is, and space to add more detailed technical
+design and planning information as we develop this information.
+
+The MPP itself does not need to describe the work, technical design, scope, and project plan in much detail.
+
+Projects can include work for major programs (such as Bedrock and Nitro), but they can focus on other areas, e.g. refactors for future capability,
+improving our testing infrastructure, testing and validation, and other engineering-oriented projects.
+-->
+<!--
+For ease of discussion in PRs, consider breaking lines after every sentence or long phrase.
+-->
+
+## What is the problem this project solves?
+_Describe the status quo, including any relevant context on the problem you're seeing that this project should solve. Who is the user you're solving for, and why do they care about this problem? Wherever possible, include pain points or problems that you've seen users experience to help motivate why solving this problem works towards top-line objectives._
+
+### Context and status quo
+
+Today on a new connection or stream, two [libp2p] nodes use the
+[multistream-select 1.0] protocol to negotiate the protocol to be used on said
+connection or stream. For example on a given TCP connection two libp2p nodes
+might negotiate the [Noise] security protocol, followed by negotiating the
+[Yamux] multiplexing protocol. On a given stream two nodes might negotiate the
+[Identify] protocol.
+
+### Problems
+
+- **Downgrade attacks** and **censorship resistance**: Given that
+  [multistream-select 1.0] negotiates a connection's security protocol
+  unencrypted and unauthenticated it is prone to [downgrade attack]s. In
+  addition, a man-in-the-middle can detect that a given connection is used
+  to carry libp2p traffic, allowing attackers to censor such connections.
+
+- **Connection establishment**: In addition to making us vulnerable to downgrade
+  attacks, negotiating the security protocol takes one round-trip in the common
+  case. On top of that negotiating a stream multiplexer (on TCP) takes another
+  round-trip.
+
+- **Data schema**: The [multistream-select 1.0] protocol is defined as a plaintext
+  protocol with no strict schema definition, making both implementation and
+  protocol evolution time consuming and error-prone. See [rust-libp2p/1795]
+  showcasing complexity for implementors and [specs/196] to showcase difficulty
+  evolving protocol.
+
+- **Bandwidth**: [multistream-select 1.0] is not as bandwidth efficient as it
+  could be. For example negotiating a protocol requires sending the protocol
```

Comment:
This is debatable and not relevant in all scenarios.

Let's put the bandwidth requirement of a plaintext protocol string in relation to the remaining bytes to be send with the example below:

Say we want to send a Kademlia `FindNode` request to a remote peer. In order to do that we would need to:

1. Negotiate the Kademlia protocol on a new stream
3. Send the `FindNode` message payload to the remote.

For (1) we will need to send `/multistream/1.0.0` (18 bytes) and `/ipfs/kad/1.0.0` (15 bytes). For (2) we would need to send the `FindNode` message with the peer ID of the node to be found (44 bytes).

With Protocol Select we would not have to send `/multistream/1.0.0` (18 bytes). In addition, with the above proposed bandwidth optimization, instead of sending `/ipfs/kad/1.0.0` (15 bytes), we might be able to reduce the protocol identifier to 2 bytes.

All that said, _Protocol Select_ will likely not tackle this in the first iteration / roll out version. It will only be designed in a way to enable this optimization in future versions.